src/main/kotlin/uk/gov/justice/digital/hmpps/externalusersapi/security/MaintainUserCheck.kt

```diff
@@ -25,31 +26,38 @@ class MaintainUserCheck(
   suspend fun ensureMaintainerGroupRelationship(
     maintainerName: String,
     groupCode: String,
-  ) {
-    // if they have maintain privileges then all good
+  ) = coroutineScope {
+
+    // All good if user holds maintain privilege
     if (authenticationFacade.hasRoles("ROLE_MAINTAIN_OAUTH_USERS")) {
-      return
+      return@coroutineScope
     }
-    val maintainer = userService.getUserAndGroupByUserName(maintainerName)
-    // otherwise group managers must have a group in common for maintenance
-    if (maintainer.groups.none { it.groupCode == groupCode }) {
-      // no group in common, so disallow
+
+    // Otherwise, group managers must have a group in common for maintenance
+    val groupsByUsername = async { groupRepository.findGroupsByUsername(maintainerName) }
+    if (groupsByUsername.await().toList().none { it.groupCode == groupCode }) {
+      // No group in common, so disallow
       throw GroupRelationshipException(groupCode, "Group not with your groups")
     }
   }
 
   @Throws(UserGroupRelationshipException::class)
-  suspend fun ensureUserLoggedInUserRelationship(loggedInUser: String, authorities: Collection<GrantedAuthority>, user: User) {
-    // if they have maintain privileges then all good
+  suspend fun ensureUserLoggedInUserRelationship(
+    loggedInUser: String,
+    authorities: Collection<GrantedAuthority>,
+    user: User,
+  ) = coroutineScope {
+
+    // All good if user holds maintain privilege
     if (canMaintainUsers(authorities)) {
-      return
+      return@coroutineScope
     }
-    // otherwise group managers must have a group in common for maintenance
-    val loggedInUserEmail = userService.getUser(loggedInUser)
-    val userGroups = groupRepository.findGroupsByUsername(user.getUserName()).toSet()
+    // Otherwise, group managers must have a group in common for maintenance
+    val loggedInUserGroups = async { groupRepository.findGroupsByUsername(loggedInUser) }
```

Comment:
To run multiple parallel Coroutine add scope 'CoroutineStart.LAZY' e.g. async(start = CoroutineStart.LAZY) {

Reply:
Just had quick look at documentation - looks like the default behaviour is what we want here i.e. immediately schedule coroutine for execution?

Reply:
Discussed & resolved